### DISCUSSION TITLE
Liquid and Nunjucks documentation, advanced examples, debugging

### DISCUSSION THREAD

**Discussion by xplosionmind** [11ty/eleventy#2275] [Q&A]
Hi everybody! I am coming from Jekyll, where I just started taking advantage of some pretty advanced Liquid tricks, when I decided to move to Eleventy.

The problem is that, while cool and useful, Liquid is terribly hard to debug. I **hate** it so much: it took me more than one year to effortlessly achieve what I intended to. The problem is that the most advanced filters I was using are [Jekyll-only ones](https://jekyllrb.com/docs/liquid/filters/). Now, I am stuck with re-learning all that I need.

Since [LiquidJS](https://liquidjs.com) and documentations are pretty limited and terribly scarce in examples, I need to figure out stuff by myself, and I do not have enough material to do it. I considered switching to [Nunjucks](https://mozilla.github.io/nunjucks/), since I sensed that Eleventy prefers it (probably because it is natively in JavaScript), but the problem is pretty much the same.

I am writing here since I would love if you could provide:
- Some more advanced (not only the basic stuff) examples with verbose explanation
- All of your ways to debug both templating languages
- Some places where I could find more extensive documentation

Thank you very much for your time!

**Comment by pdehaan** [comment]
> _"I considered switching to [Nunjucks](https://mozilla.github.io/nunjucks/), since I sensed that Eleventy prefers it (probably because it is natively in JavaScript)"_

I don't think Eleventy prefers one way or the other. LiquidJS is natively JavaScript as well (well, written in TypeScript and then transpiled to JavaScript, but the point still works).

Can you maybe expand a bit on which of the [advanced Jekyll-only filters](https://jekyllrb.com/docs/liquid/filters/) you were using?
I think most of those wouldn't be terribly difficult to convert to generic JavaScript filters which would work with Nunjucks or LiquidJS (although I would probably avoid any of the `*_exp` filters since that would be pretty challenging to convert to a generic filter that would work in all cases &mdash; and I think you'd be better off just writing specific filters for specific cases).

**Reply by xplosionmind** [reply]
> I don't think Eleventy prefers one way or the other. LiquidJS is natively JavaScript as well

What I meant is that Liquid is developed in Ruby by Shopify, then represented in JavaScript by LiquidJS. It is written in JavaScript indeed, but the original upstream is not. I am aware that Eleventy is language-agnostic, yet I see a tendency to use Nunjucks in Eleventy users.

<br>

> Can you maybe expand a bit on which of the [advanced Jekyll-only filters](https://jekyllrb.com/docs/liquid/filters/) you were using?

You would be very kind to help with specific examples, @pdehaan. Nevertheless, I hoped I could just get some links pointing to resources to learn stuff by myself. In any case, I definitely cannot refuse and I warmly welcome any more specific advice (also because in #68, #148 and  #549 my same needs emerged by other users, and it would be nice to publish a solution for each filter, once and for all).

The filters I am using the most are:
- [`absolute_url`](https://jekyllrb.com/docs/liquid/filters/#absolute-url) (currently using `{{ url | prepend: 'https://tommi.space' }}` as a workaround)
- [`where`](https://jekyllrb.com/docs/liquid/filters/#where) (currently rewriting everything to insert `if` statements inside for loops to check whether what would be the “`where` condition” is met or not)
- [`find`](https://jekyllrb.com/docs/liquid/filters/#find), specially for i18n purposes: I would set a `ref: pagename` parameter in the front matter of every page that has a duplicate in a different language, then, in the language switch, I would use `find` to find the translated page: `{{ collections.all | find: 'ref', ref }}`, where the second `ref` stands for the `ref` value of the current page.
- [`slugify`](https://jekyllrb.com/docs/liquid/filters/#slugify), mainly to manually create IDs in HTML pages. (my current workaround is to use e.g. `{{ entry.example | replace: ' ', '-' }}`, it is ineffective since it works only for whitespace characters and I would need to ad too many replace conditions to make it effective)
- [`sort`](https://jekyllrb.com/docs/liquid/filters/#sort)

Thanks a lot for your help!
T

**Reply by pdehaan** [reply]
> > I don't think Eleventy prefers one way or the other. LiquidJS is natively JavaScript as well
> 
> What I meant is that Liquid is developed in Ruby by Shopify, then represented in JavaScript by LiquidJS. It is written in JavaScript indeed, but the original upstream is not. I am aware that Eleventy is language-agnostic, yet I see a tendency to use Nunjucks in Eleventy users.

Not sure it matters (re: upstream language/origin). LiquidJS does a fantastic job of keeping things in sync w/ the Ruby implementation as much as possible. Plus, remember that per the [Nunjucks documentation](https://mozilla.github.io/nunjucks/templating.html):  _"**Nunjucks is essentially a port of [jinja2](http://jinja.pocoo.org/docs/)**, so you can read their [docs](http://jinja.pocoo.org/docs/templates/) if you find anything lacking here. Read about the differences [here](http://mozilla.github.io/nunjucks/faq.html#can-i-use-the-same-templates-between-nunjucks-and-jinja2-what-are-the-differences)."_

So:
- Liquid/Ruby 👉  LiquidJS/JavaScript
- Jinja/Python 👉  Nunjucks/JavaScript

Another templating option is also Eleventy's own https://www.11ty.dev/docs/languages/javascript/, but that means writing a lot more custom filters/logic.

**Reply by pdehaan** [reply]
> Nevertheless, I hoped I could just get some links pointing to resources to learn stuff by myself. In any case, I definitely cannot refuse and I warmly welcome any more specific advice

In what is probably the single least helpful comment you'll see today, I'd suggest probably going through the list of Jekyll filters below and seeing how many you can convert yourself into Eleventy filters (you could even create+publish your own "eleventy-plugin-jekyll-filters" module to npm) and then write up a blog post (or thirty) on lessons learned along the way. Then you can link to the plugin from the 11ty site and probably add links to your blog posts as a resources to the 11ty site and be the docs/resources you want to see. Probably learn a lot of great stuff along the way, not to mention helping future travellers and Jekyll converts (or just people looking for some great little utility filters).

**Comment by cfjedimaster** [comment]
To your third point, remember that for the template languages themselves, Eleventy shouldn't be repeating _all_ of the docs. Instead, users should go to the language's project pages for complete documentation.

**Reply by xplosionmind** [reply]
Thanks for your reply. I think I did not explain myself properly… I am writing here not because I think Eleventy should provide more information, but rather if there exist more sources of documentation, other than the official ones (they are not too helpful)

**Reply by cfjedimaster** [reply]
Do you mean more docs on Eleventy, or the languages? I'm not trying to punt the ball here, but it sounded like you wanted more info that was specific to a language, and typically I'd say start at the language docs first. If I'm misreading you, just let me know. :) Also please see @pdehaan 's comment above, if you can give some examples of what you want to do in Nunjucks we could possibly help. 

Also: Nunjucks has  a dump filter useful for debugging: https://mozilla.github.io/nunjucks/templating.html#dump
Eleventy has log (https://www.11ty.dev/docs/filters/log/)

**Comment by pdehaan** [comment]
My decidedly luke-warm takes on Jekyll's filters, and how/if I'd attempt to recreate them.

| <span style="white-space: nowrap;">[Jekyll Filter](https://jekyllrb.com/docs/liquid/filters/)</span> | NOTES
|:----|:----|
| `relative_url` | Use [`pathPrefix`](https://www.11ty.dev/docs/config/#deploy-to-a-subdirectory-with-a-path-prefix) config and [`url` filter](https://www.11ty.dev/docs/filters/url/).
| `absolute_url` | Use [`pathPrefix`](https://www.11ty.dev/docs/config/#deploy-to-a-subdirectory-with-a-path-prefix) config, [`url` filter](https://www.11ty.dev/docs/filters/url/), and write a custom filter that prefixes the domain using `new URL()`. See https://www.11ty.dev/docs/plugins/rss/#supplies-the-following-nunjucks-filters for possible examples. Or the eleventy-plugin-rss plugin's [`absoluteUrl` filter](https://github.com/11ty/eleventy-plugin-rss/blob/master/src/absoluteUrl.js) implementation.
| `date_to_xmlschema` | Custom filter using [Luxon date library](https://moment.github.io/luxon/#/formatting). See https://www.11ty.dev/docs/plugins/rss/#supplies-the-following-nunjucks-filters for possible examples.
| `date_to_rfc822` | Custom filter using [Luxon date library](https://moment.github.io/luxon/#/formatting).
| `date_to_string` | Custom filter using [Luxon date library](https://moment.github.io/luxon/#/formatting) or possibly [Date's `toLocaleDateString()` API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString).
| `date_to_long_string` | Custom filter using [Luxon date library](https://moment.github.io/luxon/#/formatting), specifically the [Presets](https://moment.github.io/luxon/#/formatting?id=presets) and [`toFormat`](https://moment.github.io/luxon/#/formatting?id=toformat) sections.
| `where` | Seems to be LiquidJS native [`where` filter](https://liquidjs.com/filters/where.html) (per v8.1.0; so available in Eleventy 1.0, but not 0.12). Might be compatible w/ the Jekyll filter. If you use Nunjucks or or another language, you'd have to write the logic yourself. [See LiquidJS's [TypeScript] implementation](https://github.com/harttle/liquidjs/blob/1af8553651277b872207e299f4d3447798c8c90d/src/builtin/filters/array.ts#L58-L66)
| `where_exp` | :shrug:
| `find` | It looks like this might be the same as the `where` filter, but just returns the first result (or nil/undefined if no item found). Could possibly use `where: "X", "Y" \| first`.
| `find_exp` | :shrug:
| `group_by` | Not sure, but possibly custom filter w/ Lodash's [`_.groupBy()`](https://lodash.com/docs/4.17.15#groupBy).
| `group_by_exp` | :shrug:
| `xml_escape` | LiquidJS's [`escape`](https://liquidjs.com/filters/escape.html) filter? Nunjucks also has an [`escape`](https://mozilla.github.io/nunjucks/templating.html#escape-aliased-as-e) filter, but I think escapes strings by default (unless changed in your config).
| `cgi_escape` | Possibly JavaScript's own native [`encodeURI()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI) or [`encodeURIComponent()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent).
| `uri_escape` | See `cgi_escape` comment above.
| `number_of_words` | Nunjucks has a builtin [`wordcount` filter](https://mozilla.github.io/nunjucks/templating.html#wordcount), but not sure how/if it would be able to support the "cjk" or "auto" arguments. You'd probably need to write your own custom filter, and even then not sure how you'd get that working; if it's a feature you need. We could probably find a way of taking a Nunjucks filter and shimming it into LiquidJS, or, it seems to be a relatively small filter, https://github.com/mozilla/nunjucks/blob/fd500902d7c88672470c87170796de52fc0f791a/nunjucks/src/filters.js#L620-L626.
| <nobr>`array_to_sentence_string`</nobr> | I'd probably use something like [`Intl.ListFormat` API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat).
| `markdownify` | Custom filter which wraps [markdown-it](https://github.com/markdown-it/markdown-it) library's `.render()` or `.renderInline()` method (depending on if you want paragraph wrap).
| `smartify` | I think [markdown-it](https://github.com/markdown-it/markdown-it) supports this via the `typographer` option, but that would probably only apply to Markdown files/blocks. If you want it elsewhere, I'd probably just create a custom filter which wraps something like https://www.npmjs.com/package/smartquotes
| `sassify` / `scssify` | Probably a custom filter which wraps https://www.npmjs.com/package/node-sass.
| `slugify` | Eleventy has built-in filters for this, although not with custom options, so if you need custom slugifying needs I'd probably just use https://www.npmjs.com/package/@sindresorhus/slugify and overwrite the built-in slugify filter.
| `jsonify` | LiquidJS has a built-in [`json`](https://liquidjs.com/filters/json.html) filter (as of v9.10.0). Otherwise, a custom filter that calls JavaScript's `JSON.stringify()` ([which is exactly what LiquidJS does](https://github.com/harttle/liquidjs/blob/1af8553651277b872207e299f4d3447798c8c90d/src/builtin/filters/misc.ts#L12-L14)). **UPDATE:** Hat tip to @cfjedimaster for reminding me of Nunjucks' [`dump` filter](https://mozilla.github.io/nunjucks/templating.html#dump).
| `normalize_whitespace` | Probably a custom filter which does a RegExp replace of multi-whitespace to a single space.
| `sort` | Both [LiquidJS](https://liquidjs.com/filters/sort.html) and [Nunjucks](https://mozilla.github.io/nunjucks/templating.html#sort-arr-reverse-casesens-attr) have `sort` filters. Not sure either support the optional `nils order`, but both have `first` and `last` filters which would accomplish the same thing (presumably, I haven't used Jekyll in a super long time, so I'm honestly guessing at most of the behaviors/implementations).
| `sample` | I'd probably create a custom filter w/ Lodash's [`_.sampleSize()`](https://lodash.com/docs/4.17.15#sampleSize).
| `to_integer` | JavaScript's built-in `parseInt()` (and also know `parseFloat()` is an option).
| `push` / `pop` / `unshift` | Not sure if LiquidJS's `append` and `prepend` would work with arrays. Otherwise probably not difficult to wrap JavaScript's native Array methods (just make sure you clone the Array first to keep it non-destructive).
| `inspect` | Use `jsonify` filter from above, or I like to wrap Node's `util.inspect()` method which handles recursive objects, sorting properties, and other tricks &mdash; but returns a non-JSON object.

**Reply by pdehaan** [reply]
It's probably a terrible idea, but that's my sweet spot&hellip;
Behold, using liquidjs filters in nunjucks (and vice-versa):

```js
const eleventyVersion = require("@11ty/eleventy/package.json").version;
const liquidFilters = require("liquidjs").filters;
const nunjucksFilters = require("nunjucks/src/filters");

module.exports = function (eleventyConfig) {
  eleventyConfig.addGlobalData("eleventyVersion", eleventyVersion);

  eleventyConfig.addFilter("wordcount", nunjucksFilters.wordcount);
  eleventyConfig.addFilter("truncatewords", liquidFilters.truncatewords);

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

```liquid
---
# src/liquid.liquid
title: The quick brown fox jumps over the lazy QA [liquidjs]
---

<h1>{{ title }}</h1>
<p>Wordcount: {{ title | wordcount }}</p>
<p>Truncate words: {{ title | truncatewords: 5 }}</p>

<p>Powered by Eleventy {{ eleventyVersion }}</p>
```

```njk
---
# src/nunjucks.njk
title: The quick brown fox jumps over the lazy QA [nunjucks]
---

<h1>{{ title }}</h1>
<p>Wordcount: {{ title | wordcount }}</p>
<p>Truncate words: {{ title | truncatewords(5) }}</p>

<p>Powered by Eleventy {{ eleventyVersion }}</p>
```

### OUTPUT
```html
<h1>The quick brown fox jumps over the lazy QA [liquidjs]</h1>
<p>Wordcount: 10</p>
<p>Truncate words: The quick brown fox jumps...</p>

<p>Powered by Eleventy 1.0.0</p>
```

---

Or, if you don't want to use the global `.addFilter()` API, you could use the more specific `.addLiquidFilter()` and `.addNunjucksFilter()`and it should all "still work":

```js
eleventyConfig.addLiquidFilter("wordcount", nunjucksFilters.wordcount);
eleventyConfig.addNunjucksFilter("truncatewords", liquidFilters.truncatewords);
```

This means that LiquidJS will use the custom "wordcount" filter, and Nunjucks would still use it's built-in "wordcount" filter.
I don't recommend mixing and matching templating languages in a project, but it's a quick hack if you want to "borrow" filters from one template engine and use them in your preferred language.

And for future reference, I was using `@11ty/eleventy@1.0.0`, `liquidjs@9.36.0`, and `nunjucks@3.2.3`.
